### ISSUE TITLE
Generation of stream names RFC

### ISSUE THREAD

**Issue by lsylvester** [rails/actioncable#26]
Currently, to stream events in a channel, you call steam_from with a string. This same string must be generated elsewhere in the application to broadcast events to this channel.

Instead of passing in a string, it could be possible to pass in `ActiveModel` compatible object and have the stream name generated from the model, and prefixed by the channel name.

So instead of:

```ruby
class WebNotificationsChannel < ApplicationCable::Channel
  stream_from "web_notifications_#{current_user.id}" 
end
```
You could have
```ruby
class WebNotificationsChannel < ApplicationCable::Channel
  stream_from current_user #=> streams from web_notifications:user_1
end
```

And then when broadcasting, instead of 
```ruby
ActionCable.server.broadcast \
  "web_notifications_1", { title: 'New things!', body: 'All shit fit for print' }
```

you could have
```ruby
WebNotificationsChannel.broadcast user, { title: 'New things!', body: 'All shit fit for print' }
  #=> broadcasts to web_notifications:user_1
```

I would be happy to make a PR for this if anyone thinks it is a good idea.

**Comment by IlyaBylich** [comment]
Sounds amazing.
But probably it would be better to have some kind of concern (`ActionCable::Support` or something like this) that will be included into `ActiveModel::Model` by default and implements some common methods like `to_actioncable_channel`? It allows to mix this module into any class (that represents a business model, but is not fully compatible with `ActiveModel::Model`) and implement this method manually.

**Comment by dhh** [comment]
I like this a lot. I'd propose a few changes. 

First, let's use the GlobalID for this. That's already a global unique identifier for a given object. We can use #to_gid_param for the identifier itself and the full identification for the logging.

Second, here's a twist to the APIs when using this type of composition:

```ruby
class WebNotificationsChannel < ApplicationCable::Channel
  def subscribed
    stream_for current_user #=> streams_from("web_notifications:Z2lkOi8vYmMzL1BlcnNvbi8x")
  end
end

WebNotificationsChannel.broadcast_to user, { title: 'New things!', body: 'All shit fit for print' }
# => WebNotificationsChannel.broadcast "web_notifications:Z2lkOi8vYmMzL1BlcnNvbi8x", ...
```

Then it's clear when we're deriving the broadcasting name vs when we're accepting the full one.

**Comment by lsylvester** [comment]
@IlyaBylich @dhh Thanks for you feedback. Using GlobalID and `stream_for` and `broadcast_to` sound like good ideas. I will have a go at implementing this.

**Comment by kyledecot** [comment]
You could shorten `when object.respond_to?(:to_global_id) then object.to_global_id.to_param` to `when object.respond_to?(:to_gid_param) then object.to_gid_param`

**Comment by kaspth** [comment]
Fixed in #34
